src/server/res.rs

```diff
@@ -9,11 +9,27 @@
 //! Response factory functions.
 //!
 
-use headers::{ContentLength, HeaderMapExt};
+use headers::{ContentLength, Header, HeaderMapExt, HeaderValue, Location};
 use hyper::StatusCode;
 
 use crate::server::Response;
 
+/// Generate 304 NotModified response.
+pub fn moved_permanently(mut res: Response, src: &str) -> Response {
+    *res.status_mut() = StatusCode::MOVED_PERMANENTLY;
+    let location = match HeaderValue::from_str(src) {
+        Ok(location) => location,
+        Err(err) => panic!("{} cannot be formatted into a Location header: {}", src, err),
+    };
+    let location = match Location::decode(&mut std::iter::once(&location)) {
+        Ok(location) => location,
+        Err(err) => panic!("{} cannot be formatted into a Location header: {}", src, err),
```

Comment:
It's not appropriate to panic directly since it's not an unexpected error. Perhaps we can return an `Result::Err` and let caller deal with it (maybe extend `internal_server_error` to show the info).

Reply:
The location string, for the use case in this PR, originates from the client request. If the location string is malformed, this probably means the client intentionally sent some characters such as `%00` (is this even possible?) to trigger this case. Perhaps `400 Bad Request` would suffice?

Reply:
If we handle the error in the `moved_permanently` caller side, we can expect that we pass in a `req.uri().path()`, which is a valid percent-encoded str. According to [RFC2616 Section-4.2], header field-content can be arbitrary `TEXT`, so passing str with percent-encoding, which is a subset of US-ASCII, is fine. 

As you mentioned, the **NUL**  is not a reserved nor unreserved character listed in [RFC3986 Section-2]. It will be encoded to `%00`, which is also a valid value that fits header's field-content. 

In conclusion, `req.uri().path()` gives us a valid percent-encoded value to pass into header. If we still get an error, it should be a problem from our side (or from headers crate 😆). So, personally 500 is fine.

[RFC2616 Section-4.2]: https://datatracker.ietf.org/doc/html/rfc2616#section-4.2
[RFC3986 Section-2]: https://datatracker.ietf.org/doc/html/rfc3986#section-2

Reply:
I got what you mean and a request without percent-encoding URI really worth a bad requset. My contention here is I trust hyper does that for us but I am not 100% sure. And I am also ok with 400. 😄